src/Microsoft.AspNet.OData.Shared/Builder/ODataInstanceAnnotationContainer.cs

```diff
@@ -0,0 +1,125 @@
+        /// </summary>
+        public ODataInstanceAnnotationContainer()
+        {
+            instanceAnnotations = new Dictionary<string, IDictionary<string, object>>();
+        }
+
+        /// <summary>
+        /// Method to Add an Instance Annotation to the CLR type
+        /// </summary>
+        /// <param name="annotationName">Name of Annotation</param>
+        /// <param name="value">Value of Annotation</param>
+        public void AddResourceAnnotation(string annotationName, object value)
+        {
+            SetInstanceAnnotation(string.Empty, annotationName, value);
+        }
+
+        /// <summary>
+        /// Method to Add an Instance Annotation to a property
+        /// </summary>
+        /// <param name="propertyName">Name of the property</param>
+        /// <param name="annotationName">Name of Annotation</param>
+        /// <param name="value">Value of Annotation</param>
+        public void AddPropertyAnnotation(string propertyName, string annotationName, object value)
+        {
+            SetInstanceAnnotation(propertyName, annotationName, value);
+        }
+
+        /// <summary>
+        /// Get an Instance Annotation from CLR Type
+        /// </summary>
+        /// <param name="annotationName">Name of Annotation</param>
+        /// <returns>Get Annotation value for the given annotation</returns>
+        public object GetResourceAnnotation(string annotationName)
+        {
+            return GetInstanceAnnotation(string.Empty, annotationName);
+        }
+
+        /// <summary>
+        /// Get an Instance Annotation from the Property
+        /// </summary>
+        /// <param name="propertyName">Name of the Property</param>
+        /// <param name="annotationName">Name of the Annotation</param>
+        /// <returns>Get Annotation value for the given annotation and property</returns>
+        public object GetPropertyAnnotation(string propertyName, string annotationName)
+        {
+            return GetInstanceAnnotation(propertyName, annotationName);
+        }
+
+        /// <summary>
+        /// Get All Annotations from CLR Type
+        /// </summary>
+        /// <returns>Dictionary of string(annotation name) and object value(annotation value)</returns>
+        public IDictionary<string, object> GetAllResourceAnnotations()
+        {
+            return GetAllInstanceAnnotations(string.Empty);
+        }
+
+        /// <summary>
+        /// Get all Annotation for a Property
+        /// </summary>
+        /// <param name="propertyName">Name of Property</param>
+        /// <returns>Dictionary of string(annotation name) and object value(annotation value)</returns>
+        public IDictionary<string, object> GetAllPropertyAnnotations(string propertyName)
+        {
+            return GetAllInstanceAnnotations(propertyName);
+        }
+
+        private void SetInstanceAnnotation(string propertyName, string annotationName, object value)
```

Comment:
>annotationName [](start = 71, length = 14)

Shall we verify duplication annotation name used?

Reply:
Duplicate will just replace (set semantics, not add semantics), which should be allowed. Right?

---
In reply to: [490460339](https://github.com/OData/WebApi/pull/2219#discussion_r490460339) [](ancestors = 490460339)

Reply:
Emm, Replace silently?  Maybe it's better to give customer an error message that you are duplicating the existing annotation? 

---
In reply to: [490467875](https://github.com/OData/WebApi/pull/2219#discussion_r490467875) [](ancestors = 490467875,490460339)

Reply:
That would force them to remove and then add? Or have separate methods for add versus set?

I think just providing the common set behavior is simpler.

---
In reply to: [490469296](https://github.com/OData/WebApi/pull/2219#discussion_r490469296) [](ancestors = 490469296,490467875,490460339)